contributors/design-proposals/kubelet-server-certificate-bootstrap-rotation.md

```diff
@@ -0,0 +1,97 @@
+# Kubelet Server Certificate Bootstrap & Rotation
+
+## Overview
+
+Currently, when a kubelet first starts, it generates a self signed
+certificate/key pair that is used for accepting incoming TLS connections. This
+proposal covers a process for generating a key locally and then issuing a
+Certificate Signing Request to the cluster API server to get an associated
+certificate signed by the cluster Certificate Authority. Also, as certificates
+approach expiration, the same mechanism will be used to request an updated
+certificate.
+
+## Use Cases
+
+1. Kubelet starts for the first time on a new node.
+2. Kubelet certificate approaches expiration.
+
+## Proposed Design
+
+### During Kubelet Boot Sequence
+
+1. Look on disk for an existing cert/key pair.
+1. If there is an existing cert/key pair, load them.
+1. If there is no existing cert/key pair, at the point where the kubelet
+   currently generates a key and certificate, the kubelet will:
+    - generate a key
+    - create a certificate signing request
+    - request a signed certificate from the API server
+    - wait for a response
+    - Store the new cert/key pairs in the directory specified by `--cert-dir`,
+      or the default value if it is unspecified.
+
+### As Expiration Approaches
+
+Rotating a Kubelet server certificate will work by generating a new private key,
+issuing a new Certificate Signing Request to the API Server, safely updating the
+cert/key pair on disk, begin using the new cert/key pair.
+ 
+1. Store the new cert/key pairs in the directory specified by `--cert-dir`, or the
+   default value if it is unspecified. This will allow the kubelet to have a
+   place for storing the multiple cert/key pairs that it might have available at
+   any given moment (because of a rotation in progress).
+    - When cert/key data is directly embedded in the kubelet config file
+      (client-certificate-data, client-key-data)
+        - The values be used only if there are no updated cert/key pairs due to
```

Comment:
this is all under a heading of serving-certificate management... serving certs are not embedded in kubeconfig files anywhere